core/trino-main/src/main/java/io/trino/sql/planner/LiteralInterpreter.java

```diff
@@ -216,4 +221,32 @@ protected Object visitNullLiteral(NullLiteral node, Void context)
             return null;
         }
     }
+
+    private static class GenericLiteralEvaluator
+    {
+        private final Type type;
+        private final boolean isJson;
+        private final ResolvedFunction resolvedFunction;
+
+        public GenericLiteralEvaluator(Type type, boolean isJson, ResolvedFunction resolvedFunction)
+        {
+            this.type = requireNonNull(type, "type is null");
+            this.isJson = isJson;
+            this.resolvedFunction = requireNonNull(resolvedFunction, "resolvedFunction is null");
+        }
+
+        Object evaluate(InterpretedFunctionInvoker functionInvoker, ConnectorSession connectorSession, GenericLiteral node)
+        {
+            Slice value = utf8Slice(node.getValue());
+            if (isJson) {
+                return functionInvoker.invoke(resolvedFunction, connectorSession, ImmutableList.of(value));
+            }
+            try {
+                return functionInvoker.invoke(resolvedFunction, connectorSession, ImmutableList.of(value));
+            }
+            catch (IllegalArgumentException e) {
+                throw semanticException(INVALID_LITERAL, node, "No literal form for type %s", type);
+            }
```

Comment:
Why do we need a separate path for json? It seems it's only to decide whether to throw an INVALID_LITERAL error or just propagate the original exception. In that case, this would make the intention clearer:

```java
try {
    return functionInvoker.invoke(...);
}
catch (...) {
    if (!isJson) {
        throw semanticException(...);
    }
}
```

Reply:
> t seems it's only to decide whether to throw an INVALID_LITERAL error or just propagate the original exception.

correct

(it was the behavior before my changes, i retained it)

> in that case, this would make the intention clearer:

i think adding a code comment would be even cleaner, will do.